Support related errors in `error-stack`

indietyp opened the pull request:
## 🌟 What is the purpose of this PR?

This extends the `error-stack` implementation to support multiple frames and sources, and also removes the `ReportImpl` in favor of an attachment-based approach.

## 🔗 Related links

* [Discussion](https://github.com/hashintel/hash/discussions/737)


## 🚫 Blocked by

* #777 

## 🏃 Follow Up

The initial PR has been split up into three parts:

1) implementation (this one)
2) hooks (new output in ASCII, fancy and json) (not yet opened)
3) #777 

## 🔍 What does this change?

* merge `ReportImpl` and `Report`
* change from single frames to frame


## 📜 Does this require a change to the docs?

Yes

* adds breaking change
* possibility of multiple sources

## ⚠️ Known issues

## 🐾 Next steps

* [x] merge `ReportImpl` and `Report`
* [x] add multiple sources to `Report` and `Frame`
* [x] common trait implementation for `Report`, this includes: `Extend`, `FromIterator`, ...
* [x] update RequestRef and co
* [x] change rendering logic
* [x] write up documentation


## 🛡 What tests cover this?

* unit tests
* integration tests for the formatting

## ❓ How to test this?

1.  Checkout the branch / view the deployment
2.  Use as a library in a binary crate
3.  Create multiple errors
4.  Merge those errors and see the output or query for information

## 📹 Demo

The output shown here has been deferred to another PR, that will be opened after this one.

<img width="907" alt="image" src="https://user-images.githubusercontent.com/7252775/177989206-4c465a36-24c8-4495-8e62-4911e73739f9.png">

<img width="241" alt="image" src="https://user-images.githubusercontent.com/7252775/177989167-113ed4a8-f3a9-4b0a-94df-01b0e7fc8e7f.png">

<img width="747" alt="image" src="https://user-images.githubusercontent.com/7252775/177989344-0c0bfa93-27ff-4c8d-90ee-6c191e41349d.png">


(without the fancy feature flag)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/hashintel/hash?pullRequest=747) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/hashintel/hash?pullRequest=747) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/hashintel/hash?pullRequest=747) it.</sub>

indietyp commented:
@TimDiekmann I am currently doing all the changes, and have kind of a problem, maybe you can help me. Because a report now holds multiple frames, I renamed `frame()` to `frames()`, but the old iterator over the stack of frames used `frames()`. I renamed that one now `traverse()` as we traverse down the tree, but I don't know if that is the right call. What do you think?

indietyp commented:
Maybe we could call it `stack()` and `stack_mut()`?

TimDiekmann reviewed (commented):
This is looking very promising so far! 🎉

indietyp commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -444,25 +453,25 @@ impl<C> Report<C> {
-    pub(crate) const fn frame(&self) -> &Frame {
-        &self.inner.frame
+    /// Returns a shared reference to the most recently added [`Frames`].
+    pub(crate) const fn frames(&self) -> &[Frame] {
```
`Frames` and `FramesMut` both use `frame()`, but they are indeed crate-local.

indietyp commented:
note: tests currently fail due to `traverse()` and `const` on slices and iter.

TimDiekmann commented:
Don't worry about the CI too much until you have finished. I just approved the PR so you can see what we are actually checking (we rely on `cargo make` so you can easily test with the same parameters as in CI). One thing worth noting is that we support `no_std` so thinks like `Vec` and `Box` must come from `alloc::`.

indietyp commented:
I wonder if it would be good to implement something similar like the old `frame()` in the public interface so that other Display and Debug implements (like hook) can also render a tree.

indietyp commented:
@TimDiekmann something that currently creates a lot of confusion is how it creates children. I am currently just taking 2 reports and then appending the frames from one to another, which can lead to surprising results, what I'd like to have instead is a NoOp frame, which has a bunch of frames as children, therefore we "bunch" them together. What do you think? How would I go about creating such a noop frame? I could in theory use `()` as attachment, but I don't know if I should.

TimDiekmann commented:
> I tried some stuff w/ the new Display (I am not done yet), but the idea would be to have something like:
> 
> <img alt="image" width="377" src="https://user-images.githubusercontent.com/7252775/177352928-c850ea80-db34-4d7d-a326-735e65b19baf.png">
> 
> (the arrows are going to be red)
> 
> My current Display render logic is a bit problematic therefore no real screenshots yet.
> 
> The snippet that should generate this in on lno. 134 in `display.rs`

I noticed you already pushed new commits. Could you resolve your issues?

indietyp commented:
(Just a Note, disregard my previous 2 comments - one deleted - because I have found a better way to represent the tree)

indietyp commented:
<img width="380" alt="image" src="https://user-images.githubusercontent.com/7252775/177536048-559e1749-a84c-44d1-94b4-d9e5e6906831.png">

This is my current version w/o a backtrace.

<img width="343" alt="image" src="https://user-images.githubusercontent.com/7252775/177536126-d3399265-0198-469a-995d-d5a46dabee0c.png">

This is my current version with a backtrace. I don't like that the backtrace is before the error.

To elaborate: My current implementation first searches along the tree and batches all single outgoing into a list. Only when we hit a "group" (multiple sources), do we indent. The first frame is the title (which has no indentation). The rest then indent. We roughly follow how Miette does this.

The backtrace is the first item since we add it as an attachment after the error. We could remedy this by first capturing the backtrace than the error, or so some more special handling for backtrace and span trace (currently, there is some special handling - but only to make them printable and defer the list to the end)

What isn't shown in the screenshots is that the backtraces are recorded and then outputted after our fancy list. I found that an indented backtrace/span trace was just very confusing.

One possibility (for reordering) is that we could do two different things:

1) first, create the backtrace frame, then the actual error frame in new
2) reorder the context frame in a display so that it is first

both have their downsides, but I currently am more in favor of (1)

The display.rs (might rename and relocate) currently only uses the public API, so other hooks can implement the same behavior if they want to, except for `report.frames`, so I'd like to propose a new method on the report: `.sources()` which gets the immediate frames and is consistent with `.add_source()`. I'd also like to add a new function called `.push()` (name TBD) to push a context as a source instead of only a report. This may make things more ergonomic.

(wow this is a lot more text than I anticipated)

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -653,6 +632,11 @@ impl<Context> std::process::Termination for Report<Context> {
     }
 }
 
+// TODO:
+// we might want this to be implemented for `Report<Context>`
+// instead, in that case tho `Report<Context>` would be empty,
+// which would break an invariant, but using `Option<Report<Context>>` can be a bit unwieldy, maybe
+// we can have an alias?
```
I think we simply don't want it yet. We could still add it at a later point if we find a good solution

nonparibus commented:
Hi @indietyp, thanks for readying this for review! Awesome work.

TimDiekmann commented on packages/libs/error-stack/.gitignore:
```diff
@@ -1,2 +1,3 @@
 target/
 Cargo.lock
+.idea
```
```suggestion

```

Alfred-Mountfield replied:
I assume you added this because you opened the error-stack folder as a root, rather than the repo? `.idea` is [already ignored in the root](https://github.com/hashintel/hash/blob/48f3b82744a91e7e934cbfc356eb1e7017b35e10/.gitignore#L79) `.gitignore`

If you're opening the error-stack folder as root, then us suggesting you remove this line might be frustrating for local development. I recommend you consider some of the approaches [outlined here](https://blog.martinhujer.cz/dont-put-idea-vscode-directories-to-projects-gitignore/).

TimDiekmann commented on packages/libs/error-stack/Cargo.toml:
```diff
@@ -17,21 +17,25 @@ tracing-error = { version = "0.2.0", optional = true }
 once_cell = { version = "1.10.0", optional = true }
 pin-project = { version = "1.0.10", optional = true }
 futures-core = { version = "0.3.21", optional = true, default-features = false }
+smallvec = { version = "1.9.0", optional = true }
```
I think we want the `union` feature for an even smaller small vec?
```suggestion
smallvec = { version = "1.9.0", optional = true, default-features = false, features = ["union"] }
```

indietyp replied:
I thought about this, but I wanted to keep this at the discretion of the user. Whenever the user uses small, I assume they use `smallvec` themselves, and if they wanted, they could toggle `union`. Otherwise, we would add `union` for them, even if they use it in a different use-case, which might be surprising if that makes sense. (at least that's what I understood from [feature unifcation](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification). Maybe creating a note in the docs would be a good idea?

TimDiekmann replied:
That's a good point. Could you add that to the documentation next to the feature flags at the top-level module docs?
I think we want to set `default-features = false` (even if it has no effect currently) anyway? (Should have been done for the other dependencies as well, which was probably an oversight, feel free to drive-by these).

TimDiekmann commented on packages/libs/error-stack/Cargo.toml:
```diff
@@ -17,21 +17,25 @@ tracing-error = { version = "0.2.0", optional = true }
 serde = { version = "1.0.137", features = ["derive"] }
 serde_json = "1.0.81"
 futures = { version = "0.3.21", default-features = false, features = ["executor"] }
+expect-test = "1.3.0"
```
Is it worth adding this as a dependency?

indietyp replied:
I have been experimenting with the integration tests but then moved to a more simplistic test. Therefore I should remove this dependency.

TimDiekmann reviewed (commented):
Thank you very much, this is really a great feature which I'm looking forward to. 🎉 

Could you please update the branch to get the latest changes from `main`?
I will review afterward 🙂

indietyp commented:
Tests fail because of the number of layers depending on the features enabled. If backtrace is enabled, we add another layer, the same if we enable span tracee. Therefore I need to rewrite some of the tests.

TimDiekmann commented on packages/libs/error-stack/src/iter.rs:
```diff
@@ -47,14 +101,14 @@ impl fmt::Debug for Frames<'_> {
     pub(crate) fn new<C>(report: &'r mut Report<C>) -> Self {
         Self {
-            current: Some(NonNull::from(report.frame_mut())),
+            current: report.frames.iter_mut().map(NonNull::from).rev().collect(),
```
We might want to store the iterator rather than a collection. This makes the traversal lazy rather then precalculating everything.
You could also apply this to `Frames` (remove the const, we have a breaking change anyway and there is no real need for a `const new`)

indietyp replied:
Yep makes sense, frames initially was the same as FramesMut but due to const limitations I needed to get creative, if it's save to remove const I'll gladly rework

indietyp replied:
I have been able to convert the `Frames` to an iterator based approach, but I am currently failing with `NonNull` and FramesMut, due to the fact that I cannot properly type the attribute, without using `dyn` (it's late in my timezone, so maybe sleeping over the whole thing might help^^).

indietyp replied:
@TimDiekmann I have been able to get a version working for `FramesMut`, but I am unsure about the soundness (I am pretty new to unsafe rust). Could you review it and make sure I didn't do anything wrong?

TimDiekmann replied:
Sure, I'll take a look, but probably not before tomorrow or Monday.

TimDiekmann replied:
Would this PR would your life make a little bit easier? 🙂 

- https://github.com/hashintel/hash/pull/774

TimDiekmann replied:
https://github.com/hashintel/hash/blob/45db0a71259d25adae5b100f460e08bda8ead133/packages/libs/error-stack/src/iter.rs#L138-L142

I'm actually not 100% sure if this is sound. My guess would be yes. The `IterMut` will hold a mutable reference to all remaining frames and you `pop` the reference from the vector in `fn next()`. I'm not sure, because the `next()` function isn't really easy to read so it's hard to reason about. In any case, the SAFETY section on the `unsafe` block is not reflecting the actual unsafe code.

A suggestion which should make this easier:
- The iterator holds frames to visit.
    ```rust
    pub struct FramesMut<'r> {
        to_visit: Vec<NonNull<Frame>>,
        _marker: PhantomData<&'r mut Frame>,
    }
    ```
- It's populated with all root frames (should probably take `&[Frame]` instead of `&[Report]`). It's pushed in reverse order so it's easier to pop a `Frame` and add new to the back:
    ```rust
    pub(crate) fn new<C>(report: &'r mut Report<C>) -> Self {
        Self {
            to_visit: report.frames.iter_mut().rev().map(NonNull::from).collect(),
            _marker: PhantomData,
        }
    }
    ````
- The `next` implementation would look like this:
    ```rust
    unsafe {
        let current = self.to_visit.last_mut()?.as_mut();
        self.to_visit
            .extend(current.sources_mut().iter_mut().rev().map(NonNull::from));
        Some(current)
    }
    ```

Do you have a test case for this? The tree shown in the documentation should probably also used for testing and then let `miri` decide, if this is sound 🙂

indietyp replied:
Yep, like the Frames implementation i tried to make it as lazy as possible, which then resulted in the usage of raw pointer. You are correct the comment doesn't reflect the actual code anymore. There are two test cases at the bottom (both Frames and FramesMut) which take the tree in the doc comment and verify that it works.

indietyp replied:
I tested the whole implementation against `miri`, and it seems like it is indeed sound. I have also rewritten the `SAFETY` comment.

If we want to have an eagerly evaluated list instead, I can roll back to the previous version, which made use of `Vec` and `NonNull`, instead of a list of iterators.

indietyp replied:
I converted this to an Iterator, instead of the `Vec<NonNull<Frame>>`, after your comment two days ago

> We might want to store the iterator rather than a collection. This makes the traversal lazy rather then precalculating everything.

The code snippet you showed me was actually pretty much the implementation before the change^^ so I can roll back if that makes it clearer.

TimDiekmann replied:
Almost, I asked you to change the inner `Vec` to an iterator 🙂
Given, that I was *almost* sure it's sound and `miri` is running fine, I'm satisfied, so you may want to leave it as it is. My solution would have been an alternative, but I don't claim that it would have been better.

Your latest commit made the `fn next()` more readable, and I like the fact it can be used for both iterators. My approach probably would have needed `unsafe` in the shared iterator as well, which is bad. The SAFETY section looks good. Only two small nits: Can you remove the `_marker` as the lifetime is now captured in the iterator? And `(&mut *frame)` can be changed to `(*frame)`.

TimDiekmann replied:
I think if we do micro-optimization, this would probably be a good spot to look at again, but without any performance baseline, it's pointless IMO. I don't see an issue in performance here currently as this is most likely the error **handling** part, not the error **creation** part.

TimDiekmann commented on packages/libs/error-stack/src/lib.rs:
```diff
@@ -78,7 +78,18 @@
 //! }
 //! # let report = read_file("test.txt").unwrap_err();
-//! # assert_eq!(report.frames().count(), 1);
+//! #
+//! # #[allow(unused_mut)]
+//! # let mut count = 1;
+//! # #[cfg(all(nightly, feature = "std"))]
+//! # {
+//! #   count += 1;
+//! # }
+//! # #[cfg(feature = "spantrace")]
+//! # {
+//! #   count += 1;
+//! # }
+//! # assert_eq!(report.frames().count(), count);
```
The tests are already pretty unreadable in code. I won't block the PR on that but we should move out the actual tests to a tests module. If you chose not to do that, could you add a `// TODO: ...` please?

indietyp replied:
Yes, I'll gladly convert them. Trying to maintain/edit the different examples is quite cumbersome.

indietyp replied:
Deferred to doctest update PR

TimDiekmann commented on packages/libs/error-stack/src/lib.rs:
````diff
@@ -198,13 +231,24 @@
 //!              at main.rs:7:10
 //! ```
 //!
+//!
````
```suggestion
```

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -205,13 +234,94 @@ impl<C> Report<C> {
+
+        #[cfg(all(nightly, feature = "std"))]
+        if let Some(backtrace) = backtrace {
+            let frame = Frame::from_attachment(backtrace, Location::caller(), source);
```
We better use the provided method `.attach(backtrace)`. This ensures, that the root frame is *always* a `Context`.

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
````diff
@@ -205,13 +234,94 @@ impl<C> Report<C> {
+    /// ```
+    ///
+    /// [`add_source()`]: Self::add_source
+    pub fn add_source<T>(&mut self, mut report: Report<T>) {
+        self.frames.append(&mut report.frames);
````
This is not adding a `source` vertically but adding a `frame` horizontally. I don't see a reason for duplicating the `extend` implementation. We however could keep it and make it non-`pub`.

indietyp replied:
that makes sense, I just saw it as a convenience, as otherwise, one might write:

```rust
use std::iter::once;

report.extend(once(other_report));
```

TimDiekmann replied:
That's a valid point. Do we want to rename it to `extend_one`? Don't have a real opinion on the name, but `source` is misleading.

indietyp replied:
I definitely agree, I think `extend_one` is a good name!

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -352,13 +440,10 @@ impl<C> Report<C> {
     pub fn span_trace(&self) -> Option<&SpanTrace> {
         #[cfg(not(nightly))]
-        let span_trace = self.inner.span_trace.as_ref()?;
+        return None;
```
Spantraces are not tight to the nightly channel. As they are an attachment, you can get them by iterating the frames. For nightly channels we should keep the `request_ref` as done below.

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -367,6 +452,15 @@ impl<C> Report<C> {
         }
     }
 
+    /// Return the direct source frames of this report,
+    /// to get an iterator over the topological sorting of all frames refer to [`frames()`]
+    ///
+    /// [`frames()`]: Self::frames
+    #[must_use]
+    pub fn sources(&self) -> &[Frame] {
+        &self.frames
+    }
+
```
This does not return the sources of the report but the current frames. This is considered the *current* error.

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -521,26 +605,7 @@ impl<Context> fmt::Display for Report<Context> {
-                write!(fmt, ": {frame}")?;
-            }
-        }
+        fmt.write_str(&report(self))?;
```
The pretty print should happen in the `Debug` implementation instead. `Display` should still be a one-liner, as `Display` is typically a user-facing string, and having line breaks in those strings often messes up the surrounding format. You probably simply want to move this call to `Debug` and slightly adjust the current `Display` implementation.

indietyp replied:
Deferred to hook update PR

TimDiekmann reviewed (commented):
Partial review, I haven't really looked at the display implementation yet. Very exciting! 🙂

TimDiekmann commented on packages/libs/error-stack/Cargo.toml:
```diff
@@ -29,11 +30,13 @@ futures = { version = "0.3.21", default-features = false, features = ["executor"
-std = ["anyhow?/std"]
+default = ["std", "small", "fancy"]
+std = ["anyhow?/std", "once_cell?/std"]
+fancy = []
```
Thinking about this a little bit more: Probably we should just keep it as the logic is already implemented. I don't really have an opinion. Ideas?

indietyp replied:
I thought what might be neat is the inclusion of display as a module and then provide several optional hooks people can use, like json, fancy, ascii, with ascii being the default.

That way people wouldn't repeat creating or copying the same code/example.

indietyp replied:
@TimDiekmann what do you think?
I'd move display.rs to `hook/fancy.rs`, move `hook.rs` to `hook/mod.rs`, and then include a json one for `hook/json.rs`. This way I could also include `termcolor` without worrying too much about a new dependency and `fancy` would still be optional.

TimDiekmann replied:
I think this would be an option. I wonder if you want to stash your changes for `fancy` (and maybe `json` if you already did some work on it) and want to open another PR for that? Then the PRs would be a little bit more self-contained.

TimDiekmann commented:
We probably should adjust the parameters in `Makefile.toml` which features to run. 192 different feature combinations are pretty hard.

indietyp commented on packages/libs/error-stack/src/display.rs:
```diff
@@ -0,0 +1,337 @@
+use crate::{AttachmentKind, Frame, FrameKind, Report};
+
+#[cfg(feature = "fancy")]
+const VERTICAL: &str = "\x1b[31m│\x1b[0m   ";
```
I didn't want to introduce a new dependency so chose to just use the [ANSI escape codes](https://en.m.wikipedia.org/wiki/ANSI_escape_code), which nowadays every terminal emulator supports.

indietyp replied:
deferred to hook PR

indietyp commented on packages/libs/error-stack/src/compat/anyhow.rs:
```diff
@@ -43,30 +44,43 @@ impl<T> IntoReportCompat for core::result::Result<T, AnyhowError> {
         match self {
             Ok(t) => Ok(t),
             Err(anyhow) => {
+                #[allow(unused_mut)]
+                let mut source: Box<[Frame]> = Box::new([]);
+
+                #[cfg(all(nightly, feature = "std"))]
+                if let Some(backtrace) = anyhow
+                    .chain()
+                    .all(|error| error.backtrace().is_none())
+                    .then(Backtrace::capture)
+                {
+                    let frame = Frame::from_attachment(backtrace, Location::caller(), source);
+                    source = Box::new([frame]);
+                }
+
+                #[cfg(feature = "spantrace")]
+                {
+                    let frame = Frame::from_attachment(
+                        tracing_error::SpanTrace::capture(),
+                        Location::caller(),
+                        source,
+                    );
+                    source = Box::new([frame]);
+                }
+
```
after some more thinking, we could also collect backtraces and spantraces and "reorder" their output (I have a rough idea as to how you could do that), making it possible to use attach() and making Context always the root.

Tho this will make the test checking for messages a bit more complicated.

indietyp commented:
> We probably should adjust the parameters in `Makefile.toml` which features to run. 192 different feature combinations are pretty hard.

Yea, with the recent adding of eyre, anyhow, small, and fancy the list of possible combinations exploded - which means running cargo make test with the current combination is very slow.

TimDiekmann commented:
As `SpanTrace` and `BackTrace` are not that special anymore (not stored inside of `Report` but "only" attached), I wonder if we still need the functions for return them. This could lead to unexpected behavior we have branches in our source-tree. I think the behavior would be the same as traversing the frames, but we now definitely have more than only one backtrace or spantrace. 
We can keep them for now, but maybe we want to deprecate them later on

indietyp commented on packages/libs/error-stack/src/display.rs:
```diff
@@ -0,0 +1,337 @@
+                | std::backtrace::BacktraceStatus::Disabled
+        ) {
+            return None;
+        }
```
Yes, well it only makes them hard in the context of doctests (because we cannot use the test_helpers crate), for this purpose I created the functions: `expect_messages` and `expect_count`, which automatically inject (if supported and needed) the `Opaque` frames in between. This makes the code clean and is pretty straightforward.

TimDiekmann replied:
It may make sense to drop the assertions in the doc tests and instead write proper tests. This is a task I'm happy to pick as a follow-up if you don't want to do the additional work 🙂

indietyp replied:
Deferred to doctest update PR

indietyp commented:
> As `SpanTrace` and `BackTrace` are not that special anymore [...]

Accessing `SpanTrace` and `Backtrace` is quite common I imagine, so we should either write something down in the documentation or maybe convert them to iterators?

TimDiekmann commented:
I think a good hint in the documentation, maybe an example, should be good enough, so we keep focusing on the actual core of the crate.

indietyp commented:
From the review (here for greater visibility) I'd be in favor of splitting this PR in 3 parts:
1) main part (this PR)
2) fancy hook (as part of also introducing a json hook)
3) doctest rework

This is per suggestion from @TimDiekmann and I'd be willing to do all of those changes, but splitting it up might be a good idea nevertheless, as this PR is already quite large.

TimDiekmann commented:
We might want to merge https://github.com/hashintel/hash/pull/777 first. From that point, it's easier to test the library. Re: doc-tests: I'm planning not running the permutated feature tests for doc tests. Also, I think `miri` isn't really required to run for doc-tests. This will greatly reduce the complexity of doc-tests.

TimDiekmann reviewed (changes requested):
PR overall looks good. Almost there! 🎉

indietyp commented on packages/libs/error-stack/src/lib.rs:
```diff
@@ -357,19 +400,26 @@
 //!  `anyhow`  | Provides conversion from [`anyhow::Error`] to [`Report`]       |         | disabled
 //!   `eyre`   | Provides conversion from [`eyre::Report`] to [`Report`]        |         | disabled
 //!
+//! You can enable further size optimizations, by including [`smallvec`] in your dependencies and
+//! enabling the `union` feature. This will reduce the size of [`Report`] by one machine word.
+//!
```
Union needs 1.49, so that's fine. I don't think that there's any downside, but I'd still feel kinda iffy to implicitly enable union for every other crate.

indietyp replied:
but maybe I am overthinking this.

TimDiekmann replied:
I think it's fine to simply use the feature 👍🏻

indietyp commented:
(just a note: running all tests takes me ~17 minutes, which is quite a lot)

TimDiekmann commented on packages/libs/error-stack/src/iter.rs:
```diff
@@ -1,22 +1,104 @@
 
 /// Iterator over the [`Frame`] stack of a [`Report`].
 ///
+/// This uses an implementation of the Pre-Order, NLR Depth-First Search algorithm to resolve the
+/// tree. The example below shows in numbers the index of the different depths, using this we're
+/// able to linearize all frames and sort topologically, meaning that this ensures no child ever is
+/// before its parent.
```
```suggestion
/// This uses an implementation of the Pre-Order, NLR Depth-First Search algorithm to resolve the
/// tree.
///
/// Use [`Report::frames()`] to create this iterator.
///
/// # Example
///
/// This shows in numbers the index of the different depths, using this it's possible to linearize
/// all frames and sort topologically, meaning that this ensures no child ever is before its parent.
```

TimDiekmann commented on packages/libs/error-stack/src/iter.rs:
````diff
@@ -1,22 +1,104 @@
+/// 3  4  5  7
+/// ```
+///
 /// Use [`Report::frames()`] to create this iterator.
````
```suggestion
```

TimDiekmann commented on packages/libs/error-stack/src/iter.rs:
```diff
@@ -1,22 +1,104 @@
+///
 /// Use [`Report::frames()`] to create this iterator.
+///
+/// We use a reversed stack of the implementation, considering the following tree:
```
```suggestion
/// A reversed stack of the implementation is used. Considering the following tree:
```

TimDiekmann commented on packages/libs/error-stack/src/iter.rs:
````diff
@@ -1,22 +1,104 @@
+/// D   E F
+/// ```
+///
+/// The algorithm will create the following through iteration:
````
```suggestion
/// the algorithm will create the following through iteration:
```

TimDiekmann commented on packages/libs/error-stack/src/lib.rs:
````diff
@@ -205,6 +202,56 @@
+//!
+//! ```rust
+//! # #![cfg_attr(not(feature = "std"), allow(dead_code, unused_variables, unused_imports))]
+//! # use std::{fs, path::Path};
+//! # use error_stack::{IntoReport, Report};
+//! # pub type Config = String;
+//!
+//! # #[cfg(all(not(miri), feature = "std"))] {
+//! fn parse_configs(paths: &[impl AsRef<Path>]) -> Result<Vec<Config>, Report<std::io::Error>> {
+//!     let mut configs = Vec::new();
+//!     let mut error: Option<Report<std::io::Error>> = None;
+//!
+//!     for path in paths {
+//!         let path = path.as_ref();
+//!
+//!         match fs::read_to_string(path).into_report() {
+//!             Ok(ok) => {
+//!                 configs.push(ok);
+//!             }
+//!             Err(err) => {
+//!                 if let Some(error) = error.as_mut() {
+//!                     error.extend_one(err);
+//!                 } else {
+//!                     error = Some(err);
+//!                 }
+//!             }
+//!         }
+//!     }
+//!
+//!     if let Some(error) = error {
+//!         return Err(error);
+//!     }
+//!
+//!     Ok(configs)
+//! }
+//!
+//! # let report = parse_configs(&["test.txt", "test2.txt", "test3.txt"]).unwrap_err();
+//! # assert!(report.contains::<std::io::Error>());
+//! # }
+//! ```
````
We don't run `miri` in doc tests anymore. Also, the doc tests are run with `--all-features`. This should simplify the examples a lot. I didn't adjust the other examples yet, but newly added examples can ignore the `cfg` dance.

indietyp replied:
True, I created these example before merging #777 and didn't adjust them yet.

TimDiekmann commented on packages/libs/error-stack/src/lib.rs:
````diff
@@ -205,6 +202,56 @@
+//! # assert!(report.contains::<std::io::Error>());
+//! # }
+//! ```
+//!
+//! [`Extend`]: std::iter::Extend
````
I think this is in the prelude?
```suggestion
```

TimDiekmann commented on packages/libs/error-stack/src/lib.rs:
```diff
@@ -357,19 +404,25 @@
 //!  `anyhow`  | Provides conversion from [`anyhow::Error`] to [`Report`]       |         | disabled
 //!   `eyre`   | Provides conversion from [`eyre::Report`] to [`Report`]        |         | disabled
 //!
+//! You can enable further size optimizations, by including [`smallvec`] in your dependencies and
+//! enabling the `union` feature. This will reduce the size of [`Report`] by one machine word.
+//!
```
```suggestion
```

TimDiekmann commented on packages/libs/error-stack/tests/common.rs:
```diff
@@ -175,3 +179,132 @@ pub fn messages<E>(report: &Report<E>) -> Vec<String> {
+
+#[cfg(all(nightly, feature = "std"))]
+pub fn supports_backtrace() -> bool {
+    // we need to track 3 states:
+    //  1) -1 (not checked)
+    //  2) 0 (disabled)
+    //  3) 1 (enabled)
+    static STATE: AtomicI8 = AtomicI8::new(-1);
+
+    match STATE.load(Ordering::SeqCst) {
+        -1 => {
+            let bt = std::backtrace::Backtrace::capture();
+            if bt.status() == std::backtrace::BacktraceStatus::Captured {
+                STATE.store(1, Ordering::SeqCst);
+                true
+            } else {
+                STATE.store(0, Ordering::SeqCst);
+                false
+            }
+        }
+        0 => false,
+        1 => true,
+        _ => unreachable!(),
+    }
```
Not blocking the PR but I think using a once-cell is easier here.

indietyp replied:
Yes, but OnceCell (and Lazy) are either unstable features or require the hooks feature. I could add `once-cell` as a dev-dependency tho. (didn't want to introduce a new dependency)

TimDiekmann replied:
As we already depend on `once-cell` it's fair to use it here as well. TBH `once-cell` is pretty simple in comparison to other dev-dependencies we use (esp. tracing is very heavy).

TimDiekmann commented on packages/libs/error-stack/tests/common.rs:
```diff
@@ -175,3 +179,132 @@ pub fn messages<E>(report: &Report<E>) -> Vec<String> {
+
+#[cfg(feature = "spantrace")]
+pub fn supports_spantrace() -> bool {
+    static STATE: AtomicI8 = AtomicI8::new(-1);
+
+    match STATE.load(Ordering::SeqCst) {
+        -1 => {
+            let st = tracing_error::SpanTrace::capture();
+            if st.status() == tracing_error::SpanTraceStatus::CAPTURED {
+                STATE.store(1, Ordering::SeqCst);
+                true
+            } else {
+                STATE.store(0, Ordering::SeqCst);
+                false
+            }
+        }
+        0 => false,
+        1 => true,
+        _ => unreachable!(),
+    }
```
Same as above

TimDiekmann commented on packages/libs/error-stack/tests/common.rs:
```diff
@@ -175,3 +179,132 @@ pub fn messages<E>(report: &Report<E>) -> Vec<String> {
+    // are both added after the `Context` therefore we need to push those layers, then `Context`
+    let last = messages.pop().unwrap();
+
+    #[cfg(all(nightly, feature = "std"))]
+    if supports_backtrace() {
+        messages.push("Opaque");
+    }
+
+    #[cfg(feature = "spantrace")]
+    if supports_spantrace() {
+        messages.push("Opaque");
+    }
```
Is this required because multiple traces are now a thing, so `zip`ping does not get rid of them?

indietyp replied:
Yep, that's precisely why. This worked before (before one of my reworks), where we would attach the backtrace and spantrace before the context. But, because now those are added before the `Context,` we have a problem because now we have two conditional Opaque layers just before the root layer and can no longer easily use `zip`.

This is why `expect_messages`, `expect_count` and `assert_kinds!` exists.

Example:
The actual layer is:
`["Printable A", "Printable B", "Opaque"?, "Opaque"?, "Context"]` where the Opaque layer is only added if backtraces are enabled and can be captured. I didn't want to repeat the logic (shown above) every single time, so I create functions to accommodate for that issue.

TimDiekmann replied:
That makes sense, thanks for the explanation!

TimDiekmann commented on packages/libs/error-stack/tests/common.rs:
```diff
@@ -175,3 +179,132 @@ pub fn messages<E>(report: &Report<E>) -> Vec<String> {
+}
+
+#[allow(unused_macros)]
+macro_rules! assert_kinds {
```
Could you add a comment on what this macro is expecting/doing?

TimDiekmann commented on packages/libs/error-stack/tests/test_backtrace.rs:
```diff
@@ -14,7 +14,11 @@ fn captured() {
 
     let report = create_report();
-    let backtrace = report.backtrace().expect("No backtrace captured");
+    assert_eq!(report.request_ref::<std::backtrace::Backtrace>().count(), 1);
```
You maybe want to import `Backtrace` into scope 😉

TimDiekmann commented on packages/libs/error-stack/tests/test_compatibility.rs:
```diff
@@ -121,12 +161,17 @@ fn anyhow_output() {
-    ONCE.call_once(|| {
-        eyre::set_hook(Box::new(eyre::DefaultHandler::default_with)).expect("Could not set hook");
-    })
+    static ONCE: core::sync::atomic::AtomicBool = core::sync::atomic::AtomicBool::new(false);
+
+    if ONCE
+        .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        .is_ok()
+    {
+        eyre::set_hook(alloc::boxed::Box::new(eyre::DefaultHandler::default_with))
+            .expect("Could not set hook");
+    }
```
What was wrong with using `std::sync::Once`?

TimDiekmann replied:
Can we revert it then back to `Once`? In tests, `std` is always available, even if the feature is not specified. However, gated features in the library won't be available without the `std` feature. You can see `std` as an implicit dev-dependency.

indietyp replied:
Yes, already did in one of my upcoming commits.

TimDiekmann commented on packages/libs/error-stack/tests/test_compatibility.rs:
```diff
@@ -121,12 +161,17 @@ fn anyhow_output() {
 }
 
-#[cfg(feature = "eyre")]
+#[cfg(all(feature = "eyre"))]
```
```suggestion
#[cfg(feature = "eyre")]
```

TimDiekmann commented on packages/libs/error-stack/tests/test_debug.rs:
```diff
@@ -43,7 +43,7 @@ Caused by:
     );
 
-    assert!(format!("{report:?}").starts_with(&expected_output));
+    // assert!(format!("{report:?}").starts_with(&expected_output));
```
Can we `#[ignore]` the test instead?

TimDiekmann commented on packages/libs/error-stack/tests/test_debug.rs:
```diff
@@ -113,5 +113,5 @@ fn extended() {
     );
 
-    assert!(dbg!(format!("{report:#?}")).starts_with(&dbg!(expected_output)));
+    // assert!(dbg!(format!("{report:#?}")).starts_with(&dbg!(expected_output)));
```
Same as above

TimDiekmann commented on packages/libs/error-stack/src/frame/mod.rs:
```diff
@@ -31,24 +31,38 @@ impl Frame {
     #[must_use]
-    pub fn source(&self) -> Option<&Self> {
-        self.frame.source()
+    pub fn sources(&self) -> &[Self] {
+        self.frame.sources()
```
We never check more than one element in tests

TimDiekmann commented on packages/libs/error-stack/src/frame/mod.rs:
```diff
@@ -31,24 +31,38 @@ impl Frame {
     #[must_use]
-    pub fn source_mut(&mut self) -> Option<&mut Self> {
-        self.frame.source_mut()
+    pub fn sources_mut(&mut self) -> &mut [Self] {
+        self.frame.sources_mut()
```
We never check more than one element in tests

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -27,26 +30,40 @@ use crate::{
 ///
+/// ## Multiple Errors
+///
+/// `Report` is able to represent multiple errors that have occurred.
+/// Errors can be combined using the [`extend_one()`],
+/// which will add the [`Frame`] stack of the other error as an additional source to the current
+/// report.
```
```suggestion
/// `Report` is able to represent multiple errors that have occurred. Errors can be combined using
/// the [`extend_one()`], which will add the [`Frame`] stack of the other error as an additional
/// source to the current report.
```

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -27,26 +30,40 @@ use crate::{
-/// [`Backtrace`] or a [`SpanTrace`], those are returned, otherwise, if configured, an attempt is
-/// made to capture them when creating a `Report`. To enable capturing of the backtrace, make sure
-/// `RUST_BACKTRACE` or `RUST_LIB_BACKTRACE` is set according to the
+/// calling [`request_ref::<Backtrace>()`] ([`downcast_ref::<Backtrace>()`] on stable) or
```
`Backtrace` is not available on `stable`

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
````diff
@@ -138,33 +155,67 @@ use crate::{
+///
+/// Get the attached backtrace and spantrace:
+///
+/// ```
+/// # #![cfg_attr(nightly, feature(backtrace))]
+///
+/// # #[cfg(all(not(miri), feature = "std"))] {
+/// use error_stack::{IntoReport, ResultExt, Result};
+///
+/// # #[allow(dead_code)]
+/// # fn fake_main() -> Result<String, std::io::Error> {
+/// let config_path = "./path/to/config.file";
+/// let content = std::fs::read_to_string(config_path)
+///     .into_report()
+///     .attach_printable_lazy(|| format!("Failed to read config file {config_path:?}"));
+///
+/// let content = match content {
+///     Err(err) => {
+///         # #[cfg(all(nightly, feature = "std"))]
+///         for backtrace in err.request_ref::<std::backtrace::Backtrace>() {
+///             println!("Backtrace: {backtrace}");
+///         }   
+///
+///         # #[cfg(all(nightly, feature = "spantrace"))]
+///         for spantrace in err.request_ref::<tracing_error::SpanTrace>() {
+///             println!("Spantrace: {spantrace}")
+///         }
+///
+///         return Err(err)
+///     }
+///
+///     Ok(ok) => ok
+/// };
+///
+/// # const _: &str = stringify! {
+/// ...
+/// # }; Ok(content) }
+/// # }
+/// ```
````
Same as above re: `miri` and `--all-features` in doc tests. This, however, does not apply to `stable` vs `nightly`.

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -211,13 +262,112 @@ impl<C> Report<C> {
+        this
+    }
+
+    #[allow(missing_docs)]
+    #[must_use]
+    #[cfg(all(nightly, feature = "std"))]
+    #[deprecated = "a report might contain multiple backtraces, use `request_ref::<Backtrace>()` \
+                    instead"]
+    pub fn backtrace(&self) -> Option<&Backtrace> {
+        self.request_ref::<Backtrace>().next()
+    }
```
Function is not covered in tests

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -211,13 +262,112 @@ impl<C> Report<C> {
+        self.request_ref::<Backtrace>().next()
+    }
+
+    #[allow(missing_docs)]
+    #[must_use]
+    #[cfg(feature = "spantrace")]
+    #[cfg_attr(
+        nightly,
+        deprecated = "a report might contain multiple spantraces, use \
+                      `request_ref::<SpanTrace>()` instead"
+    )]
+    #[cfg_attr(
+        not(nightly),
+        deprecated = "a report might contain multiple spantraces, use \
+                      `frames().filter(Frame::downcast_ref::<SpanTrace>)` instead"
+    )]
+    pub fn span_trace(&self) -> Option<&SpanTrace> {
+        #[cfg(nightly)]
+        return self.request_ref::<SpanTrace>().next();
+
+        #[cfg(not(nightly))]
+        return self.downcast_ref::<SpanTrace>();
+    }
```
Function is not covered in tests

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -211,13 +262,112 @@ impl<C> Report<C> {
+    /// };
+    ///
+    /// use error_stack::{Context, Report};
+    /// # #[cfg(all(not(miri), feature = "std"))]
```
```suggestion
```

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -314,87 +452,44 @@ impl<C> Report<C> {
-        } else {
-            None
-        }
+    pub fn current(&self) -> &[Frame] {
```
Maybe rename to `current_frames`?

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -314,87 +452,44 @@ impl<C> Report<C> {
-    /// Returns the span trace of the error, if captured.
-    ///
-    /// Note, that [`ErrorLayer`] has to be enabled to enable span traces.
+    /// Return the direct source frames of this report,
+    /// to get an iterator over the topological sorting of all frames refer to [`frames()`]
```
Could you please mention the difference between this and `current_context`?

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -633,10 +727,23 @@ impl<Context> std::process::Termination for Report<Context> {
-    backtrace: Option<Backtrace>,
-    #[cfg(feature = "spantrace")]
-    span_trace: Option<SpanTrace>,
+impl<Context> FromIterator<Report<Context>> for Option<Report<Context>> {
+    fn from_iter<T: IntoIterator<Item = Report<Context>>>(iter: T) -> Self {
+        let mut iter = iter.into_iter();
+
+        let mut base = iter.next()?;
+        for rest in iter {
+            base.extend_one(rest);
+        }
+
+        Some(base)
+    }
+}
```
This is not covered in tests

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
````diff
@@ -211,13 +262,112 @@ impl<C> Report<C> {
+    /// ```
+    ///
+    /// [`extend_one()`]: Self::extend_one
+    pub fn extend_one<T>(&mut self, mut report: Report<T>) {
````
When `Extend::extend_one` is stabilized, is it safe to remove this function?

indietyp replied:
Yes, that is the plan, I will add a note about the issue in the docs to make sure we keep track of it.

TimDiekmann replied:
Great, thank you! We typically add a `// TODO: ...` as well, so it's easy to find them later on.

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
````diff
@@ -211,13 +262,112 @@ impl<C> Report<C> {
+    /// ```
+    ///
+    /// [`extend_one()`]: Self::extend_one
+    pub fn extend_one<T>(&mut self, mut report: Report<T>) {
````
Let's not allow extending from a `Report` with a different `Context`.
```suggestion
    pub fn extend_one(&mut self, mut report: Self) {
```

TimDiekmann commented on packages/libs/error-stack/tests/test_compatibility.rs:
```diff
@@ -141,17 +186,40 @@ fn eyre() {
+    let eyre_report = eyre.into_report().unwrap_err();
+    let mut eyre_messages = messages(&eyre_report);
+
+    if swap {
+        // we're reversing the whole thing, but that also means that the optional opaque layer
+        // isn't at the correct place when looking at the messages.
+        // ["Root error", "Printable A", "Opaque", "Printable B"]
+        // ["Printable B", "Opaque", "Printable A", "Root error"]
+        // which isn't correct as opaque needs to be before `Root` to be represented correctly.
+
+        eyre_messages.swap(1, 2);
+    }
```
This is failing if `RUST_BACKTRACE` is not set

TimDiekmann replied:
This is still failing:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"Printable A"`,
 right: `"Printable B"`', tests/test_compatibility.rs:223:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -633,10 +727,23 @@ impl<Context> std::process::Termination for Report<Context> {
+    }
+}
+
+impl<Context> Extend<Self> for Report<Context> {
+    fn extend<T: IntoIterator<Item = Self>>(&mut self, iter: T) {
+        for item in iter {
+            self.extend_one(item);
+        }
+    }
```
Very sad we can't implement `Extend<Frame>` without allowing a `Report` to be extended from any `Report` (without matching context). Then this would have been more intuitive. 😞

indietyp replied:
Definitely 😞

TimDiekmann commented:
> (just a note: running all tests takes me ~17 minutes, which is quite a lot, if everything is compiled that number goes down to ~ 70 seconds)

Some timings for me: 
- clean build with all features: 9s
- clean build tests: 26s
- tests build; `cargo make test`: 13m 43s
- Running `cargo make test` again: 1m 14s

I did not set any environment but the CI also runs with `CARGO_INCREMENTAL=0`, which should slightly improve the timing.

The difference to your measurements are most probably due to the used hardware.

It's possible to tune the timings by setting `CARGO_CLIPPY_HACK_FLAGS` and `CARGO_TEST_HACK_FLAGS` in `Makefile.toml` to include `--group-features`, so some features will be tested at once.

TimDiekmann commented:
> 2. fancy hook (as part of also introducing a json hook)

I wonder if implementing `Serialize` for `Report` maybe makes more sense, then it's not limited to JSON.

TimDiekmann commented:
Yes, that is what I had in mind! 👍🏻 
I think we don't even need to pull `serde_derive` for this.

TimDiekmann commented on packages/libs/error-stack/src/report.rs:
```diff
@@ -211,13 +259,115 @@ impl<C> Report<C> {
+    ///
+    /// This function implements the same functionality as
+    /// [`Extend::extend_one` (#7261)](https://github.com/rust-lang/rust/issues/72631).
+    /// Once stabilised this function will be deprecated in favor of [`Extend`].
```
I guess it will be removed?
```suggestion
    /// Once stabilised this function will be removed in favor of [`Extend`].
```

indietyp commented:
I got an email about it still failing, but haven't been able to find the corresponding open review, did I miss something?

TimDiekmann reviewed (approved):
Thank you very much, great work! 🎉

indietyp commented:
🎉
